DOCSP-39702 Standardize change streams page (#69)
mongodb/docs-java-rs: source/read/change-streams.txt

.. _java-rs-change-streams:

====================
Monitor Data Changes
====================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference
 
.. meta::
   :keywords: code examples, watch

Overview
--------

In this guide, you can learn how to use a **change stream** to monitor real-time
changes to your data. A change stream is a {+mdb-server+} feature that
allows your application to subscribe to data changes on a collection, database,
or deployment.

Sample Data
~~~~~~~~~~~

The examples in this guide use the ``sample_restaurants.restaurants`` collection
from the :atlas:`Atlas sample datasets </sample-data>`. To learn how to create a
free MongoDB Atlas cluster and load the sample datasets, see the
:ref:`<java-rs-getting-started>` guide.

.. include:: includes/reactor-note.rst

Open a Change Stream
--------------------

To open a change stream, call the ``watch()`` method. The instance on which you
call the method determines the scope of events that the change

> stream listens for. You can call the ``watch()`` method on the following
> classes:

Reviewer: S: 

```suggestion
stream listens for. You can call the ``watch()`` method on instances of the following
classes:
```

- ``MongoClient``: To monitor all changes in the MongoDB deployment
- ``MongoDatabase``: To monitor changes in all collections in the database
- ``MongoCollection``: To monitor changes in the collection

The following example opens a change stream on the ``restaurants`` collection
and outputs the changes as they occur:

.. literalinclude:: /includes/read/change-streams.java
   :start-after: start-open-change-stream
   :end-before: end-open-change-stream
   :language: java

To begin watching for changes, run the application. Then, in a separate

> application or shell, modify the ``restaurants`` collection. Updating a document

Reviewer: ```suggestion
application or shell, perform a write operation on the ``restaurants`` collection. Updating a document
```

> that has a ``"name"`` value of ``"Blarney Castle"`` results in the following

Reviewer: Nit:

```suggestion
that has a ``name`` field value of ``"Blarney Castle"`` results in the following
```

Castle"`` results in the following change stream output:

.. code-block:: sh
   :copyable: false

   Received change: ChangeStreamDocument{ operationType=update, resumeToken={"_data": "..."}, 
   namespace=sample_restaurants.restaurants, destinationNamespace=null, fullDocument=null, 
   fullDocumentBeforeChange=null, documentKey={"_id": {"$oid": "..."}}, clusterTime=Timestamp{...}, 
   updateDescription=UpdateDescription{removedFields=[], updatedFields={"cuisine": "Traditional Irish"}, 
   truncatedArrays=[], disambiguatedPaths=null}, txnNumber=null, lsid=null, 
   splitEvent=null, wallTime=BsonDateTime{value=...}}

Modify the Change Stream Output
-------------------------------

You can pass an aggregation pipeline as a parameter to the ``watch()``
method to modify the change stream output. This parameter allows you to watch
for only specified change events.

You can specify the following aggregation stages in the ``pipeline`` parameter:

- ``$addFields``
- ``$match``
- ``$project``
- ``$replaceRoot``
- ``$replaceWith``
- ``$redact``
- ``$set``

> - ``$unset``

Author: For now it will have to be kept in sync manually, but there is a longer term plan in progress for the docs that will bring these repos together, which should get rid of the need to manually sync it.
